src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+(defpackage #:coalton-impl/codegen/constant-propagation
+  (:use
+   #:cl
+   #:coalton-impl/codegen/ast)
+  (:local-nicknames
+   (#:tc #:coalton-impl/typechecker))
+  (:import-from
+   #:coalton-impl/codegen/traverse
+   #:action
+   #:count-applications
+   #:count-nodes
+   #:make-traverse-let-action-skipping-cons-bindings
+   #:*traverse*
+   #:traverse
+   #:traverse-with-binding-list)
+  (:export
+   #:propagate-constants))
+
+(in-package #:coalton-impl/codegen/constant-propagation)
+
+(defvar *constant-bindings*)
+(setf (documentation '*constant-bindings* 'variable)
+      "Bound inside CONSTANT-PROPAGATION.")
+
+(defun constant-var-value (var &key no-error)
+  (the (or node null)
+       (or (cdr (assoc var *constant-bindings*))
+           (unless no-error
+             (error "Expected VAR ~S to be a constant but is not" var)))))
+
+(defun propagate-constants (node env)
+  (declare (optimize debug))
+  (labels ((constant-node-p (node)
```

Comment:
why don't we lift this out as a toplevel function

Reply:
> why don't we lift this out as a toplevel function

I assume you are referring to the `constant-node-p` function. With the suggestion that [`traverse` allows passing arbitrary arguments](https://github.com/coalton-lang/coalton/pull/1212#discussion_r1717172995), I think if we pass the `env` argument,  we can lift out `constant-node-p` as a toplevel function. Let me know if that sounds good.